### PR TITLE
Use jQuery.Deferred rather than Promise for cross-browser compatability

### PR DESCRIPTION
**Pull request by carlio**
This PR is to enable the Chat XBlock to work in Internet Explorer versions 10 and 11 - to fix this ticket: https://tasks.opencraft.com/browse/OC-2758

The incompatibility was caused by using JavaScript Promises which are not available in IE 11 and earlier. The solution is to use jQuery.Deferred instead. This was [already being used](https://github.com/open-craft/xblock-chat/blob/master/chat/public/js/src/chat.js#L433) in this XBlock so will not introduce problems.

*I have tested that this works with:*

* Standard Chrome browser
* Using IE11 on a Windows VM
* iOS simulator

*To Test*

* Use the Chat XBlock exactly as before - no behaviour has changed
* Verify that it continues to work on Firefox/Chrome etc.
* Verify that it can work on IE10 and IE11. This might require using browsertesting.com for example.

*Some notes*

* jQuery Deferred objects are not 100% equivalent to Promise objects prior to jQuery 1.9 - see https://thewayofcode.wordpress.com/tag/jquery-deferred-broken/ . However a newer version of jQuery is used.
* ~~One of the unit tests is failing, which I am investigating. That's why this is marked as 'WIP'~~ Fixed now!

*Screenshot of chat working under IE*

virtualbox_windoze_01_07_2017_16_28_47 copy

### CODE REVIEW

File: chat/public/js/src/chat.js [hunk 1]
```diff
@@ -361,9 +361,8 @@ function ChatXBlock(runtime, element, init_data) {
      * pause: delays execution with a timeout
      */
     var pause = function(timeout) {
-        return new Promise(function(resolve, reject) {
-            setTimeout(resolve, timeout);
-        });
+        var promise = $.Deferred();
+        setTimeout(promise.resolve, timeout);
```
Comment (mtyaka) [inline chat/public/js/src/chat.js:368]: You have to `return` the promise here. I think that's what's causing your tests to fail.

Comment (carlio) [reply chat/public/js/src/chat.js:368]: Ah nice catch, thanks! I tried changing that, the unit test still fails I'm getting closer to fixing that.

Comment (mtyaka) [reply chat/public/js/src/chat.js:368]: Try cherry-picking this commit from my PR: https://github.com/open-craft/xblock-chat/pull/12/commits/c2a320c7c5f5c8fa03ab63b254835e7252b1a190

That could fix the failing test.

Comment (carlio) [reply chat/public/js/src/chat.js:368]: @mtyaka Brilliant! It did, at least locally - let's see what Circle says. That saved me a big bug hunt 👍

File: chat/public/js/src/chat.js [hunk 2]
```diff
@@ -361,9 +361,9 @@ function ChatXBlock(runtime, element, init_data) {
...
+        return promise
```
Comment (mtyaka) [inline chat/public/js/src/chat.js]: nit: missing semicolon

**Review by mtyaka — COMMENTED** [review]
@carlio Are you still working on this? If not, can you please update the title and description to remove "WIP" and squash the commits (except the cherry-picked one)?

File: chat/public/js/src/chat.js (hunk 2 above)
Comment (carlio) [reply chat/public/js/src/chat.js]: It is done code-wise, I will fix the PR description and I will fix the semicolon also

**Comment by mtyaka** [comment]
@carlio BTW I tested this in IE 11 and can confirm it's working now :)

I noticed a UI bug with the message bubble tails being misplaced (also visible in your screenshot). They are positioned around 10px lower than in other browsers for some reason. I spent around 15 minutes trying to fix that, but couldn't find anything that works. Unless you find a way to fix this quickly, we can leave this for now and fix it later in a separate story.

File: chat/public/js/src/chat.js (hunk 2 above)
Comment (carlio) [reply chat/public/js/src/chat.js]: And everything is updated now

**Comment by carlio** [comment]
@mtyaka I think the UI bug bug should be a separate story but hopefully not a big one

**Review by mtyaka — APPROVED** [review]

**Comment by mtyaka** [comment]
@carlio Thanks, this looks good! Can you please squash all commits (except the cherry-picked one) and then we can merge this?

:+1:

- [x] I tested this: I interacted with the chat in IE 11, Chrome, FF, and Safari.
- [x] I read through the code
- [x] I checked for accessibility issues
- [ ] ~~Includes documentation~~ bug fix; not required

**Comment by mtyaka** [comment]
@carlio I just merged https://github.com/open-craft/xblock-chat/pull/12, which includes the commit you cherry-picked into this PR. Can you please remove the chery-picked commit, squash all other commits into a single commit, and rebase on top of master? Then we can merge this.

**Comment by carlio** [comment]
@mtyaka  Will do. I did the rebase but a unit test is failing so I'll fix that up first.

File: chat/public/js/src/chat.js (hunk 1 above)
Comment (mtyaka) [reply chat/public/js/src/chat.js:368]: @carlio It looks like the `return promise;` disappeared after rebase? I think that's what's causing the test to fail - if it doesn't return the promise then the functions using `pause` will execute immediately and not wait for that `setTimeout`.

Comment (carlio) [reply chat/public/js/src/chat.js:368]: @mtyaka The `return` is still there, it's just the line below this comment chain. So unfortunately that's not the problem here! It is something to do with the `var submitResponse` changes I made - if I replace it with the code on master all the tests pass.

Comment (mtyaka) [reply chat/public/js/src/chat.js:368]: Ah you're right, sorry. I'm not sure which branch I was looking at.

**Comment by mtyaka** [comment]
Rebased from 0afa1bedc09d66e5589258166d87eca6e0895436.

**Comment by mtyaka** [comment]
@carlio Updating `xblock-sdk` to the latest version fixed the issue with the test. I squashed some commits, rebased on top of master, and tested in common browsers including IE 11 to verify it works. I'm merging this now.

The block now loads in IE 10, but the layout is broken. I assume that's because IE 10 implements an old version of the flexbox spec. Fixing this will have to be a separate issue.

**Pull request merged**